tietokanta/src/main/resources/db/migration/V1_2170__.sql

```diff
@@ -0,0 +1,138 @@
+    edellinen_maara NUMERIC,
+    maaramuutos NUMERIC,
+    uusi_maara NUMERIC
+);
+
+
+CREATE TABLE mhu_muutos_liite (
+    versio INTEGER DEFAULT 1, -- jokainen tallennus tallentaa täyden version muutoksesta, ml. alitaulut joissa on tietoa
+    muutos INTEGER REFERENCES mhu_muutos(id),
+    liite INTEGER REFERENCES liite(id)
+);
+
+CREATE INDEX mhu_muutos_id_versio_idx ON mhu_muutos (id, versio);
+CREATE INDEX mhu_muutos_kustannusvaikutus_idx ON mhu_muutos_kustannusvaikutus (muutos, versio, hoitokauden_alkuvuosi);
+CREATE INDEX mhu_muutos_tehtava_ja_maaraluettelo_idx ON mhu_muutos_tehtava_ja_maaraluettelo (muutos, versio, hoitokauden_alkuvuosi);
+CREATE INDEX mhu_muutos_liite_idx ON mhu_muutos_liite (muutos, versio);
+
+-- historiataulujen luonti, hyödyntää INHERITS-toimintoa. INHERIT ei peri rajoitteita, joten historiataulussa
+-- voi olla mhu_muutoksia monta riviä samalla id:lla kuten halutaankin
+CREATE TABLE mhu_muutos_historia () INHERITS (mhu_muutos);
+-- varmistetaan että samaa (id, versio)-yhdistelmää on vain yksi
+CREATE UNIQUE INDEX mhu_muutos_historia_id_versio ON mhu_muutos_historia (id, versio);
+CREATE TABLE mhu_muutos_liite_historia () INHERITS (mhu_muutos_liite);
+CREATE TABLE mhu_muutos_kustannusvaikutus_historia () INHERITS (mhu_muutos_kustannusvaikutus);
+CREATE TABLE mhu_muutos_tehtava_ja_maaraluettelo_historia () INHERITS (mhu_muutos_tehtava_ja_maaraluettelo);
+
+
+-- historiatauluja päivittävät triggerit ja niiden funktiot
+CREATE FUNCTION paivita_mhu_muutos_historia()
+    RETURNS TRIGGER AS $$
+BEGIN
+    INSERT INTO mhu_muutos_historia
+    VALUES (OLD.*);
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER mhu_muutos_historia_trigger
+    BEFORE UPDATE OR DELETE ON mhu_muutos
+    FOR EACH ROW
+EXECUTE FUNCTION paivita_mhu_muutos_historia();
+
+
+CREATE FUNCTION paivita_mhu_muutos_kustannusvaikutus_historia()
+    RETURNS TRIGGER AS $$
+BEGIN
+    INSERT INTO mhu_muutos_kustannusvaikutus_historia
+    VALUES (OLD.*);
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER paivita_mhu_muutos_kustannusvaikutus_historia_trigger
+    BEFORE UPDATE OR DELETE ON mhu_muutos_kustannusvaikutus
+    FOR EACH ROW
+EXECUTE FUNCTION paivita_mhu_muutos_kustannusvaikutus_historia();
+
+
+CREATE FUNCTION paivita_mhu_muutos_tehtava_ja_maaraluettelo_historia()
+    RETURNS TRIGGER AS $$
+BEGIN
+    INSERT INTO mhu_muutos_tehtava_ja_maaraluettelo_historia
+    VALUES (OLD.*);
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER mhu_muutos_tehtava_ja_maaraluettelo_historia_trigger
+    BEFORE UPDATE OR DELETE ON mhu_muutos_tehtava_ja_maaraluettelo
+    FOR EACH ROW
+EXECUTE FUNCTION paivita_mhu_muutos_tehtava_ja_maaraluettelo_historia();
+
+CREATE FUNCTION paivita_mhu_muutos_liite_historia()
+    RETURNS TRIGGER AS $$
+BEGIN
+    INSERT INTO mhu_muutos_liite_historia
+    VALUES (OLD.*);
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER mhu_muutos_historia_trigger
```

Comment:
Tämän pitäisi olla varmaan nimeltään:
```suggestion
CREATE TRIGGER mhu_muutos_liite_historia_trigger
```

mhu_muutos_historia_trigger on määritelty tuolla ylempänä.